### PR TITLE
DOLFINX wrapper for ufc_expression

### PR DESCRIPTION
**Pull request by jhale**
Work in progress. Largely a duplication of `fem::Form` so far.

### CODE REVIEW

**Comment by michalhabera** [comment]
All looking good. Can you add Expression to the python layer as well (would be good if we have some way of testing this PR)?

In Python layer, I imagine `Expression` will be similar to `Form`, but will not inherit from any UFL parent - we do not want to allow Expression stand within a Form. So it would basically be ~
```python
class Expression:
    __init__(self, ufl_expression, reference_points):
        # call ffcx jit for ufl expression
        # ...
        # create C++ expression
``` 

One issue to discuss is the name. Currently, dolfinx c++ objects reflect the UFC names, `ufc_form --> Form`, `ufc_expression --> Expression`. This means we will have plenty of users using new Expression as a legacy Expression. But these are completely different. New one is **compiled** and cannot be put into Form unless interpolated into a Function.

Would be good to hear from @garth-wells @chrisrichardson what they think.

**Comment by garth-wells** [comment]
It's not clear to me yet what `fem::Expression` does. It doesn't have any methods that do something yet. When it's clearer how it can/should be used I think it will be clearer what an informative name would will be.

File: python/test/unit/fem/test_expression_evaluation.py [hunk 1]
```diff
@@ -97,3 +93,66 @@ def grad_expr1(x):
...
     assert np.isclose((b2.vector - b.vector).norm(), 0.0)
+
+
+def test_simple_evaluation():
+    """Test evaluation of UFL Expression.
+
+    This test evaluates a UFL Expression on two cells in the mesh and compares the
+    result with an analytical expression.
+
+    For a donor function f(x, y) = (x^2 + 2*y^2) result is compared with the
```
Comment (garth-wells) [inline python/test/unit/fem/test_expression_evaluation.py:104]: What is a 'donor' function?

File: python/test/unit/fem/test_expression_evaluation.py [hunk 2]
```diff
@@ -97,3 +93,66 @@ def grad_expr1(x):
...
+    exact gradient grad f(x, y) = [2*x, 4*y]. In this test, f(x, y) is first
```
Comment (garth-wells) [inline python/test/unit/fem/test_expression_evaluation.py:105]: I would put the equation a separate line - it's not so easy to parse the 'grad' with the 'f(x, y)'.

File: python/test/unit/fem/test_expression_evaluation.py [hunk 3]
```diff
@@ -97,3 +93,66 @@ def grad_expr1(x):
...
+    mesh = dolfinx.generation.UnitSquareMesh(MPI.COMM_WORLD, 1, 1)
+    P2 = dolfinx.FunctionSpace(mesh, ("P", 2))
+
+    f = dolfinx.Function(P2)
+
+    def exact_expr(x):
+        return x[0] ** 2 + 2.0 * x[1] ** 2
+
+    def exact_grad_expr(x):
+        values = np.empty((2, x.shape[1]))
```
Comment (garth-wells) [inline python/test/unit/fem/test_expression_evaluation.py:120]: Does `return np.array([ 2*x[0], 4*x[1] ] )` work?

File: cpp/dolfinx/function/utils.h [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  Eigen::Array<T, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> c(
+      num_cells, offsets.back());
+  if (coefficients.size() > 0)
+  {
+    for (int cell = 0; cell < num_cells; ++cell)
+    {
+      for (std::size_t coeff = 0; coeff < dofmaps.size(); ++coeff)
+      {
+        auto dofs = dofmaps[coeff]->cell_dofs(cell);
+        const Eigen::Ref<const Eigen::Matrix<T, Eigen::Dynamic, 1>>& _v
```
[the remark this answers is not in the document]
Comment (jhale) [reply cpp/dolfinx/function/utils.h:65]: I took this code verbatim from elsewhere (it usually takes fem::Form). I could fix this issue. Regarding the code duplication, which I'd also like to fix, these functions could be templated?

File: cpp/dolfinx/function/evaluate.h [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      coordinate_dofs(num_dofs_g, gdim);
+
+  Eigen::Array<T, Eigen::Dynamic, 1> values_e;
+  const Eigen::Index num_points = e.num_points();
+  const Eigen::Index value_size = e.value_size();
+  const Eigen::Index size = num_points * value_size;
+  values_e.setZero(size);
+
+  // Iterate over cells and 'assemble' into values
+  for (std::size_t i = 0; i < active_cells.size(); ++i)
```
[the remark this answers is not in the document]
Comment (jhale) [reply cpp/dolfinx/function/evaluate.h:73]: I need both `i` (the index into the array of cells) and `c` (the cell).

File: cpp/dolfinx/function/evaluate.h [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+      coordinate_dofs.row(j) = x_g.row(x_dofs[j]).head(gdim);
+
+    auto coeff_cell = coeffs.row(c);
+
+    values_e.setZero();
+    fn(values_e.data(), coeff_cell.data(), constant_values.data(),
+       coordinate_dofs.data());
+
+    for (Eigen::Index k = 0; k < size; ++k)
+      values[i * size + k] = values_e[k];
```
[the remark this answers is not in the document]
Comment (jhale) [reply cpp/dolfinx/function/evaluate.h:88]: Yes.

File: python/test/unit/fem/test_expression_evaluation.py (hunk 1 above)
Comment (michalhabera) [reply python/test/unit/fem/test_expression_evaluation.py:104]: An input function to the evaluation. The expression represents pointwise gradient() operator, so it maps input functions to output function. In different terminology, it maps a donor function to a target function. It is a donor in a sense that is provides expansion coefficients to the kernel. We can change it to "input function" as well.

File: python/dolfinx/wrappers/function.cpp [hunk 7]
```diff
@@ -134,8 +134,18 @@ void function(py::module& m)
   py::class_<dolfinx::function::Expression<PetscScalar>,
              std::shared_ptr<dolfinx::function::Expression<PetscScalar>>>(
       m, "Expression", "An Expression")
-      .def(py::init<>(),
-           "Create an Expression")
+      .def(py::init<>(), "Create an Expression")
+      .def("eval", &dolfinx::function::Expression<PetscScalar>::eval)
+      .def_property("mesh", &dolfinx::function::Expression<PetscScalar>::mesh,
```
[the remark this answers is not in the document]
Comment (jhale) [reply python/dolfinx/wrappers/function.cpp:139]: Perhaps I missed a style guide here.  Yes, `num_points` should indeed be read-only. For the others, they need to be set from the Python wrapper layer and are low cost. So, I made them read/write properties.

[the remark this answers is not in the document]
Comment (garth-wells) [reply python/dolfinx/wrappers/function.cpp:139]: The pybind11 docs are quite light in this area. Use `def_property_readonly` if it's just a getter.

**Comment by jhale** [comment]
> `Eigen::Ref` is a bit different to `&`. `Eigen::Ref<const Eigen::Matrix<T, Eigen::Dynamic, 1>` can hold a reference, for example, to a `Eigen::Matrix<T, 10, 1>`. It's probably more appropriate that `std::reference_wrapper` in this particular case.

Yep I used std::reference_wrapper, makes sense.

Two outstanding issues?

* Should Expression actually be in the `fem::` namespace? I think so, because this type of Expression is tightly related to the concept of a finite elements (we need coefficients, basis functions etc. to compute it). Any opinions?
* Can I template `pack_coefficients` and `pack_constants`? They act on Form and Expression in exactly the same way. Would reduce code duplication.

**Comment by garth-wells** [comment]
I fixed a typo in my comment. It should have said that `Eigen::Ref` is more appropriate in this case.

`std::reference_wrapper` is good where in the past we would have based a vector of pointers but where the pointers cannot be `nullptr`.

**Comment by jhale** [comment]
OK. I'll change back to `Eigen::Ref` but I'll wait until we decide to template `pack_constants` and `pack_coefficients` and move everything I've made to `fem::`.

**Review by garth-wells — CHANGES_REQUESTED** [review]
Looking forward to using this!

**Comment by jhale** [comment]
> Looking forward to using this!

So am I! There is still an outstanding issue regarding striding of vector-valued expressions that I will discuss with @michalhabera tomorrow. At the moment I require an axis swap and a copy at https://github.com/FEniCS/dolfinx/pull/1083/files#diff-5f20ff6b108855336a268f3360fe5162R236 before the evaluation can be 'naturally' inserted/assembled into a vector-valued Quadrature space.

**Comment by jhale** [comment]
I have an issue: FFCX currently outputs evaluation code for vector spaces which is not naturally compatible with the DOF ordering in DOLFINX. Specifically, it outputs XXXYYYZZZ format, rather than XYZXYZXYZ.

This requires a reordering of the data before it can be 'naturally' inserted into a VectorFunctionSpace, see e.g.

https://github.com/FEniCS/dolfinx/pull/1083/files#diff-5f20ff6b108855336a268f3360fe5162R236

This issue cannot be fixed by simply reshaping the array (while keeping a contiguous memory layout).

I would like to make the appropriate change in FFCX. Could @mscroggs point me in the right direction? I know you worked on this change elsewhere.

**Comment by mscroggs** [comment]
I think something along the lines of https://github.com/FEniCS/ffcx/compare/mscroggs/xyzxyz is what you're looking for

**Comment by jhale** [comment]
Thanks @mscroggs ! I thought you would be able to do it quickly. I'll test it with my DOLFINX branch and report back here.

**Comment by mscroggs** [comment]
Those changes are merged into FFCx master now, so you can revert back to using master in the tests

File: python/dolfinx/wrappers/function.cpp [hunk 8]
```diff
@@ -129,5 +130,37 @@ void function(py::module& m)
...
+      const Eigen::Ref<const
+          Eigen::Array<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>&,
+      const std::size_t>(), "Create an Expression")
+      .def("eval", &dolfinx::function::Expression<PetscScalar>::eval)
+      .def_property_readonly("mesh", &dolfinx::function::Expression<PetscScalar>::mesh)
+      .def_property_readonly("num_points",
+                    &dolfinx::function::Expression<PetscScalar>::num_points)
+      .def_property_readonly("value_size",
+                    &dolfinx::function::Expression<PetscScalar>::value_size)
+      .def_property_readonly("x", &dolfinx::function::Expression<PetscScalar>::x)
```
[the remark this answers is not in the document]
Comment (jhale) [reply python/dolfinx/wrappers/function.cpp:148]: The default policy for properties is `return_value_policy::reference_internal` which ensures that the parent C++ object won't go out of scope until the Python object goes out of scope. Do you want me to make it explicit?

https://pybind11.readthedocs.io/en/stable/advanced/functions.html

[the remark this answers is not in the document]
Comment (garth-wells) [reply python/dolfinx/wrappers/function.cpp:148]: Might be helpful to make it explicit - from pybind11 docs it's the default for setters/getters, whereas as the default for methods is different. A bit too subtle perhaps.

File: cpp/dolfinx/function/Expression.h [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+          coefficients)
+  {
+    for (const auto& c : coefficients)
+      _coefficients.set(c.first, c.second);
+  }
+
+  /// Set coefficient with given name (shared pointer version)
+  /// @param[in] coefficients Map from coefficient name to the
+  ///   coefficient
+  void set_coefficients(
```
[the remark this answers is not in the document]
Comment (jhale) [reply cpp/dolfinx/function/Expression.h:127]: Going to leave this one to work out a good strategy for Form. Keen to get this merged in now before it gets stale.

**Comment by jhale** [comment]
@garth-wells I think this is in good shape to merge now, not withstanding the issue of the unsafe construction. I think we should discuss this in the issue you opened and get this merged in now.

**Comment by jhale** [comment]
This has been updated to follow the best practice in `fem::Form`.

Only two remaining issues:

1) Would Expression be better off in `fem::`? It seems more like a `Form` than anything in the `function` namespace.
2) I would like to make the template type inference work, avoiding having to manually specify `T` and `U` when calling `pack_constants` and `pack_coefficients`.

**Pull request merged**